cmd/snap-confine/mount-support-nvidia.c

```diff
@@ -345,55 +343,62 @@ static void sc_mount_nvidia_driver_biarch(const char *rootfs_dir)
 
 #ifdef NVIDIA_MULTIARCH
 
-struct sc_nvidia_driver {
-	int major_version;
-	int minor_version;
-};
+typedef struct {
+	int major;
+	// Driver version format is MAJOR.MINOR[.MICRO] but we only care about the
+	// major version and the full version string. The micro component has been
+	// seen with relevant leading zeros (e.g. "440.48.02").
+	char raw[sizeof("1234.5678.90AB\n")];
+} sc_nv_version;
 
-static void sc_probe_nvidia_driver(struct sc_nvidia_driver *driver)
+static void sc_probe_nvidia_driver(sc_nv_version * version)
 {
+	memset(version, 0, sizeof *version);
+
 	FILE *file SC_CLEANUP(sc_cleanup_file) = NULL;
 	debug("opening file describing nvidia driver version");
 	file = fopen(SC_NVIDIA_DRIVER_VERSION_FILE, "rt");
 	if (file == NULL) {
 		if (errno == ENOENT) {
 			debug("nvidia driver version file doesn't exist");
-			driver->major_version = 0;
-			driver->minor_version = 0;
 			return;
 		}
 		die("cannot open file describing nvidia driver version");
 	}
-	// Driver version format is MAJOR.MINOR where both MAJOR and MINOR are
-	// integers. We can use sscanf to parse this data.
-	if (fscanf
-	    (file, "%d.%d", &driver->major_version,
-	     &driver->minor_version) != 2) {
-		die("cannot parse nvidia driver version string");
+	int nread = fread(version->raw, 1, sizeof version->raw - 1, file);
+	if (nread < 0) {
+		die("cannot read nvidia driver version string");
+	}
+	if (nread == sizeof version->raw - 1 && !feof(file)) {
+		die("cannot fit entire nvidia driver version string");
```

Comment:
Would it make sense to make the buffer a bit bigger just in case? They may start adding git revision or something.

Reply:
Yeah, we can make it something larger than a short version without being crazy. Say PATH_MAX?

Reply:
No strong opinion, PATH_MAX is 4096 iirc, fine for me but even ~128 is probably fine

Reply:
+1 I'll adjust this.